contracts/SP3QSAJQ4EA8WXEDSRRKMZZ29NH91VZ6C5X88FGZQ.crashpunks-v1.clar

```diff
@@ -0,0 +1,1004 @@
+            (currentBidder (unwrap! (get-current-bidder nftIndex currentBidIndex) bidding-error))
+            (currentAmount (unwrap! (get-current-bid-amount nftIndex currentBidIndex) bidding-error))
+            (seriesOriginal  (unwrap! (get series-original (map-get? nft-data {nft-index: nftIndex})) not-allowed))
+            (owner           (unwrap! (nft-get-owner? crashpunks nftIndex) not-allowed))
+        )
+        (asserts! (or (is-eq closeType u1) (is-eq closeType u2)) failed-to-close-1)
+        ;; only the owner or administrator can call close
+        (asserts! (or (unwrap! (is-approved nftIndex owner) nft-not-owned-err) (unwrap! (is-administrator) not-allowed)) not-allowed)
+        ;; only the administrator can call close BEFORE the end time - note we use the less accurate
+        ;; but fool proof block time here to prevent owner/client code jerry mandering the close function
+        (asserts! (or (> block-time bidding-end-time) (unwrap! (is-administrator) failed-to-close-3)) failed-to-close-3)
+
+        ;; Check for a current bid - if none then just reset the sale data to not selling
+        (if (is-eq currentAmount u0)
+            (map-set nft-sale-data { nft-index: nftIndex } { sale-cycle-index: (+ saleCycleIndex u1), sale-type: u0, increment-stx: u0, reserve-stx: u0, amount-stx: u0, bidding-end-time: u0})
+            (if (is-eq closeType u1)
+                (begin
+                    ;; buy now closure - pay and transfer ownership
+                    ;; note that the money to pay with is in the contract!
+                    (print {evt: "close-bidding", nftIndex: nftIndex, payType: "from-contract", txSender: tx-sender, currentBidder: currentBidder, currentAmount: currentAmount, currentBidIndex: currentBidIndex})
+                    (unwrap! (payment-split nftIndex currentAmount (as-contract tx-sender) seriesOriginal) payment-error)
+                    (unwrap! (nft-transfer? crashpunks nftIndex (unwrap! (nft-get-owner? crashpunks nftIndex) nft-not-owned-err) tx-sender) transfer-error)
+                )
+                (begin
+                    ;; refund closure - refund the bid and reset sale data
+                    (print {evt: "close-bidding", nftIndex: nftIndex, payType: "refund", txSender: tx-sender, currentBidder: currentBidder, currentAmount: currentAmount})
+                    (unwrap! (refund-bid nftIndex currentBidder currentAmount) failed-refund)
+                    (map-set nft-sale-data { nft-index: nftIndex } { sale-cycle-index: (+ saleCycleIndex u1), sale-type: u0, increment-stx: u0, reserve-stx: u0, amount-stx: u0, bidding-end-time: u0})
+                )
+            )
+        )
+        (print {evt: "close-bidding", nftIndex: nftIndex, closeType: closeType, txSender: tx-sender, currentBidder: currentBidder, currentAmount: currentAmount})
+        (ok nftIndex)
+    )
+)
+
+;; read only methods
```

Comment:
most read-only functions do not need a result of type `response`, better to use plain value if not required by traits. 

Specifying read-only in traits might change with Stacks 2.1: https://github.com/blockstack/stacks-blockchain/issues/1981